proposals/0000-implicit-resource-binding.md

```diff
@@ -0,0 +1,369 @@
+void main() {
+  A[0] = s.E[2][0] + s.C[1][0] + s.B[3][0] + s.E[1][0];
+}
+```
+https://godbolt.org/z/45j8aqTaf
+
+This seems wrong. Resource arrays should be bound to a continuous description
+range and the range should be reflected in the `createHandleFromBinding`
+arguments. It also makes the binding susceptible to change whenever the code
+changes and is not something that can be relied upon. 
+
+Array resources inside structs are also not allowed to use dynamic indexing. In
+the following example DXC reports error when the resource array `B` is accessed
+while the dynamic indexing of `A` with is fine.
+
+#### Example 2.3
+
+```c++
+RWBuffer<float> A[10];
+
+struct S {
+  RWBuffer<float> B[10];
+} s;
+
+[numthreads(4,1,1)]
+void main() {
+  for (int i = 0; i < 5; i++) {
+    A[i][0] = 1.0;
+    s.B[i][0] = 1.0; // error: Index for resource array inside cbuffer must be a literal expression
+  }
+}
+```
+
+### Unbound Resource Arrays
+
+Unbound resource arrays are placed in `space0` after the highest explicitly
+assigned descriptor, or after the highest implictly assigned descriptor so far,
+whichever is greater. They take up the rest of the descriptor slots in `space0`.
+
+#### Example 3.1
+
+```c++
+RWBuffer<float> A : register(u1);
+RWBuffer<float> B[];     // gets u6 (unbounded range)
+RWBuffer<float> C : register(u5);
+RWBuffer<float> D[3];    // gets u2 because it fits between A and C but not before A
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[2][0] + C[0] + B[10][0];
+}
+```
+https://godbolt.org/z/vbvfqjo8h
+
+If there are resources declared after the unbound array that do not fit into the
+remaining space available in `space0`, DXC reports an error. For example
+increasing the dimension of array 'D' from 3 to 4 results in a `error: resource
+D could not be allocated`.
+
+#### Example 3.2
+```c++
+RWBuffer<float> A : register(u1);
+RWBuffer<float> B[];     // gets u6 (unbounded range)
+RWBuffer<float> C : register(u5);
+RWBuffer<float> D[4];    // error - does not fit in the remaining descriptor ranges in space0 
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[2][0] + C[0] + B[10][0];
+}
+```
+https://godbolt.org/z/6bYoG599P
+
+It looks like DXC never attempts to assign descriptors from virtual register
+space other than `space0`.
+
+However, moving declaration of `D` before the unbounded array `B` compiles
+successfully and `B` gets assigned descriptors from `u10` forward.
+
+#### Example 3.3
+
+```c++
+RWBuffer<float> A : register(u1);
+RWBuffer<float> D[4];    // gets u6
+RWBuffer<float> B[];     // gets u10 (unbounded range)
+RWBuffer<float> C : register(u5);
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[2][0] + C[0] + B[10][0];
+}
+```
+https://godbolt.org/z/cjP35n63d
+
+Since DXC seems to only assign descriptors from `space0`, it reports an error
+whenever there are:
+- two or more unbound resource arrays without explicit binding
+- one unbound resource array with explicit binding in `space0` and one or more
```

Comment:
Note: it also emits an error with 2 explicitly bound unbounded resource arrays.
So perhaps this could be simplified to:
`two or more unbounded resource arrays in the same register space.`